### PR TITLE
Added an option to specify target framework

### PR DESCRIPTION
**Pull request by psfinaki**
Okay so 1.0 is out and I want to address the biggest pain we have :D 

This implementation works but I haven't added tests yet as I want to verify if this the right direction and if this fits the new options approach.

image

If we merge this I will go try Stryker 1.0 in action!

### CODE REVIEW

**Comment by rouke-broersma** [comment]
Do you think it will be something that can change run to run? If so a cli option is appropriate. If not there should only be a file based option.

I don't really know if people will want to run against both target frameworks or if they will generally pick one which will remain static (and should thus be in the config file).

I think you can best answer that question.

**Comment by rouke-broersma** [comment]
Btw we released an rc for v1 but not v1 yet :p

**Comment by psfinaki** [comment]
> Do you think it will be something that can change run to run? If so a cli option is appropriate. If not there should only be a file based option.

Hmm hard to say for sure but we at least will be running everything against .NET Core for now to simplify things so I imagine we'll specify this in the config. So probably we can start with that and maybe add the CLI option later.

We can also revise the whole thing once the support for runs against multiple frameworks is added - which should be cool but will need more thinking and effort to implement. We have already encountered one platform-specific mutation (for the same code) so there's potential in that topic :)

So I guess I'll remove the cli option and add tests & docs?

> Btw we released an rc for v1 but not v1 yet :p

Yeah but it's v1.0 code running and that's what matters :D

**Review by richardwerkman — CHANGES_REQUESTED** [review]
Thanks for your contribution :)

File: src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+        protected override string Description => "The framework to build the project against.";
+
+        public string Validate()
+        {
+            if (SuppliedInput is not null)
+            {
+                if (string.IsNullOrWhiteSpace(SuppliedInput))
+                {
+                    throw new InputException("Target framework cannot be empty. Please provide a valid value from this list: https://docs.microsoft.com/en-us/dotnet/standard/frameworks");
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs:17]: Well I thought about it but the list is quite long and there also are weird things there like `net6.0-ios14.0` and so on. Plus it's not very futureproof anyway... I haven't found examples for a proper approach around as well.

File: src/Stryker.CLI/Stryker.CLI/CommandLineConfigHandler.cs [hunk 2]
```diff
@@ -119,6 +119,7 @@ private void PrepareCliOptions(IStrykerInputs inputs)
             AddCliInput(inputs.DevModeInput, "dev-mode", null, optionType: CommandOptionType.NoValue);
             AddCliInput(inputs.ConcurrencyInput, "concurrency", "c", argumentHint: "number");
             AddCliInput(inputs.SolutionInput, "solution", "s", argumentHint: "file-path");
+            AddCliInput(inputs.TargetFrameworkInput, "target-framework", "t", CommandOptionType.SingleValue);
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.CLI/Stryker.CLI/CommandLineConfigHandler.cs:122]: Will do this and add tests, hang on

[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.CLI/Stryker.CLI/CommandLineConfigHandler.cs]: Right, so this is moved

File: src/Stryker.Core/Stryker.Core/Initialisation/ProjectFileReader.cs [hunk 3]
```diff
@@ -99,5 +100,28 @@ private void LogAnalyzerResult(IAnalyzerResult analyzerResult)
...
         }
+
+        private IAnalyzerResult SelectAnalyzerResult(IAnalyzerResults analyzerResults, string targetFramework)
+        {
+            if (targetFramework == null)
+            {
+                return analyzerResults.First();
+            }
+
+            var analyzerResultForFramework = analyzerResults.SingleOrDefault(result => result.TargetFramework == targetFramework);
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Initialisation/ProjectFileReader.cs:110]: Easier said than done... But it's there :) Not the most beautiful way probably. I started refactoring things a bit in the DI direction but got into a rabbit hole so stopped at the current implementation.

File: src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+using Stryker.Core.Exceptions;
+
+namespace Stryker.Core.Options.Inputs
+{
+    public class TargetFrameworkInput : Input<string>
```
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs:5]: Added tests.

File: src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (psfinaki) [reply src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs:17]: Reopen if you have any ideas.

**Comment by psfinaki** [comment]
@richardwerkman I hope I've addressed the points, please rereview :)

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            analyzerResultFrameworkYMock.Setup(m => m.TargetFramework).Returns("Y");
+
+            _analyzerResults = new[]
+            {
+                analyzerResultFrameworkXMock.Object,
+                analyzerResultFrameworkYMock.Object
+            };
+
+            var result = _projectFileReader.AnalyzeProject(null, null, null);
+            Assert.Equal("X", result.TargetFramework);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs:52]:
```suggestion
            result.TargetFramework.ShouldBe("X");
```

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            analyzerResultFrameworkYMock.Setup(m => m.TargetFramework).Returns("Y");
+
+            _analyzerResults = new[]
+            {
+                analyzerResultFrameworkXMock.Object,
+                analyzerResultFrameworkYMock.Object
+            };
+
+            var result = _projectFileReader.AnalyzeProject(null, null, "Y");
+            Assert.Equal("Y", result.TargetFramework);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs:71]:
```suggestion
            result.TargetFramework.ShouldBe("Y");
```

File: src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            analyzerResultFrameworkYMock.Setup(m => m.TargetFramework).Returns("Y");
+
+            _analyzerResults = new[]
+            {
+                analyzerResultFrameworkXMock.Object,
+                analyzerResultFrameworkYMock.Object
+            };
+
+            var result = _projectFileReader.AnalyzeProject(null, null, "Z");
+            Assert.Equal("X", result.TargetFramework);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TargetFrameworkResolutionTests.cs:90]:
```suggestion
            result.TargetFramework.ShouldBe("X");
```

File: src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardwerkman) [reply src/Stryker.Core/Stryker.Core/Options/Inputs/TargetFrameworkInput.cs:17]: We could use some sort of regex to check it. But it's very error prone indeed. For now let's keep it like this.

**Review by richardwerkman — APPROVED** [review]
Looking good! Found only some minor shouldly improvements still.

**Comment by psfinaki** [comment]
Okay that was a pleasure to fix :D

**Pull request merged**